### PR TITLE
Implement AOS frame support

### PR DESCRIPTION
**Pull request by kmarwah**
Add support for AOS frames, including the requested changes from @aywaldron

### CODE REVIEW

File: ait/dsn/sle/common.py [hunk 1]
```diff
@@ -97,6 +97,8 @@ class SLE(object):
 
     def __init__(self, *args, **kwargs):
         ''''''
+        self. _downlink_frame_type = ait.config.get('dsn.sle.downlink_frame_type',
+                                      kwargs.get('telemetryframetype', 'TMTransFrame'))
```
Comment (aywaldron) [inline ait/dsn/sle/common.py]: The kwargs key to check for here would be `downlink_frame_type` as well, not `telemetryframetype`.

File: ait/dsn/bin/ait_sle_bridge.py [hunk 2]
```diff
@@ -65,16 +65,17 @@ def process_pdu(raf_mngr):
 
 
 if __name__ == '__main__':
-    raf_mngr = ait.dsn.sle.RAF(hostname='atb-ocio-sspsim.jpl.nasa.gov', port=5100,
+    #put in if statement for each IP address
+    raf_mngr = ait.dsn.sle.RAF(hostname='137.228.89.105', port=5214, #ip address subject to change
```
Comment (aywaldron) [inline ait/dsn/bin/ait_sle_bridge.py]: If this hostname, port, and inst_id is project-specific information it should not be included in the PR.

File: ait/dsn/bin/examples/cltu_api_test.py [hunk 3]
```diff
@@ -38,9 +38,9 @@
 import ait.dsn.sle
 
 cltu_mngr = ait.dsn.sle.CLTU(
-    hostname='atb-ocio-sspsim.jpl.nasa.gov',
+    hostname='127.0.0.1', #test for now
     port=5100,
-    inst_id='sagr=LSE-SSC.spack=Test.fsl-fg=1.cltu=cltu1',
+    inst_id='sagr=39.spack=default.fsl-fg=1.cltu=cltu1',
```
Comment (aywaldron) [inline ait/dsn/bin/examples/cltu_api_test.py]: Same comment as above.

Comment (kmarwah) [reply ait/dsn/bin/examples/cltu_api_test.py]: the original "LSE-SSC" is also project specific, but from another project. Should I add another field in config.yaml for CLTU_inst_id?

File: ait/dsn/bin/ait_sle_bridge.py (hunk 2 above)
Comment (kmarwah) [reply ait/dsn/bin/ait_sle_bridge.py]: The original hostname (atb-ocio-sspsim.jpl.nasa.gov) is also project specific. Should I add a sle_hostname field in config.yaml?

**Comment by kmarwah** [comment]
@aywaldron I removed all the project specific information I could, including some references to SSPSim that are currently in the NASA-AMMOS:master branch. I also added a few more fields in config.yaml, which means that less Python code has to be edited when instantiating a CLTU/RAF/RCF class. The only thing that I didn't include in config.yaml was inst_id, because this value changes depending on whether it's for FCLTU, RAF-online, RCF-offline, etc.

File: ait/dsn/bin/ait_sle_bridge.py [hunk 4]
```diff
@@ -65,16 +65,16 @@ def process_pdu(raf_mngr):
 
 
 if __name__ == '__main__':
-    raf_mngr = ait.dsn.sle.RAF(hostname='atb-ocio-sspsim.jpl.nasa.gov', port=5100,
-                             auth_level="bind",
-                             inst_id="sagr=LSE-SSC.spack=Test.rsl-fg=1.raf=onlc1")
+    raf_mngr = ait.dsn.sle.RAF(hostname=self._hostname, port=self._port,
+                             auth_level=self._auth_level,
```
Comment (aywaldron) [inline ait/dsn/bin/ait_sle_bridge.py]: This code is not in a class so `self.` isn't going to work.

File: ait/dsn/bin/examples/cltu_api_test.py [hunk 5]
```diff
@@ -38,10 +38,10 @@
 import ait.dsn.sle
 
 cltu_mngr = ait.dsn.sle.CLTU(
-    hostname='atb-ocio-sspsim.jpl.nasa.gov',
-    port=5100,
-    inst_id='sagr=LSE-SSC.spack=Test.fsl-fg=1.cltu=cltu1',
-    auth_level="bind"
+    hostname=self._hostname,
+    port=self._port,
```
Comment (aywaldron) [inline ait/dsn/bin/examples/cltu_api_test.py]: Same comment as above.

File: ait/dsn/bin/ait_sle_bridge.py (hunk 4 above)
Comment (kmarwah) [reply ait/dsn/bin/ait_sle_bridge.py]: Good catch. I removed all the redundant arguments which are already present in config.yaml and inherited from the SLE base class defined in common.py (hostname, port, auth_level). The only required argument when instantiating a RAF/RCF/CLTU class is inst_id, which isn't defined in config.yaml because it's specific to the object being instantiated.

**Comment by aywaldron** [comment]
Hi @kmarwah, thanks for updating again. Looking at the AOS transfer frame spec at https://public.ccsds.org/Pubs/732x0b3e1.pdf, page 4-2, there are a few discrepancies between your implementation and the spec. For example, in the spec the signaling field contains 4 distinct values (it is not unused), and the last 2 octets are the optional "frame header error control", not the first header pointer. Which document are you referencing in your implementation? If your implementation has some project-specific aspects to it, I will want to edit it to make it more general after merging your PR.

**Comment by kmarwah** [comment]
hey @aywaldron, I referenced the Iris radio's ICD when building the class, so the implementation is specific to LunaH-Map and other projects using Iris (Icecube).

image

**Comment by aywaldron** [comment]
@kmarwah Ok, thanks for including that reference here. I will merge this PR and leave the ticket open for us to implement the specs in the CCSDS documentation rather than the Iris ICD.
